### ISSUE TITLE
macos - when I keep multiple databases open, tab order changes sometimes when I close then reopen application

### ISSUE THREAD

**Issue by jonathanmorgan** [keepassxreboot/keepassxc#1294]
When I keep multiple databases open, tab order changes when I close, then reopen the Keepassxc application.

## Expected Behavior
I expect the tabs to maintain their order consistently between executions of the application.

## Current Behavior
When I keep multiple databases open, tab order changes when I close, then reopen the Keepassxc application.  I try to keep them arranged so that my main database is on the right, and so is focused on first when I launch the application.  Unfortunately, the tab order changes every time I close, then re-open the application.  Sometimes the main database's tab stays to the right, so not all tabs change their position every time.

## Steps to Reproduce (for bugs)

1. Open multiple database files at once.
2. Close, then re-open the application.
3. Some or all of the tabs should have changed order.

## Context

I try to keep them arranged so that my main database is on the right, and so is focused on first when I launch the application so I can just type my pass phrase and go.  Unfortunately, the tab order changes every time I close, then re-open the application, so I've learned now that I have to look for the main database's tab every time I open the application.

## Debug Info

KeePassXC - Version 2.2.4
Revision: 4723f66

Libraries:
- Qt 5.9.1
- libgcrypt 1.8.1

Operating system: macOS High Sierra (10.13)
CPU architecture: x86_64
Kernel: darwin 17.3.0

Enabled extensions:
- KeePassHTTP
- Auto-Type
- YubiKey
Libraries:
- LIBS

**Comment by droidmonkey** [comment]
This is a feature of the app, previously open databases are opened in alphanumerical sorted order when started. I personally agree with you that the exact tab order should be restored as well.

**Comment by jonathanmorgan** [comment]
Interesting.  I’d expect it to stay consistent, then.  The thing I’ve noticed is the tabs on the end flipping.  I thought it might just be inverting the order sometimes, but hen when I was trying to see if it was consistent, it looked like it was rearranging randomly within the tab set (I usually have 5 tabs open).  Now that I know there is a sort there, i’ll Look and see if it is obvious what is happening based on the order.

**Comment by jonathanmorgan** [comment]
In looking again, as long as I don't change tab order, it looks like they are in a set order, not alphabetical order, and then they flip from being laid out left to right and right to left sometimes when I close and re-open the program.

Right now, for example, the titles start with "cu", "kl", "CDF_", "CDF.", and "ce".  They either sort in this order, or backward: "ce", "CDF.", "CDF_", "kl", "cu".  I can attach screenshots if that would help.  It is consistently this order, though.

If I adjust the order and close the program, then reopen, my new order doesn't persist, but it also does not sort either alphabetically or to the same order as before.  For example, I put them in alphabetical order ("CDF.", "CDF_", "ce", "cu", "kl") and closed the program, then re-opened it.  Now the order is "CDF.", "CDF_", "cu", "kl", and then "ce".

From this brief look, I see no pattern I can identify quickly.

**Comment by TheZ3ro** [comment]
We already fixed this one point in time, maybe the is some problem with that fix. I have to search for it in previous PRs

**Comment by droidmonkey** [comment]
OK I was completely wrong about the alphnumeric sorting... however the manner in which KeePassX saves the open tabs is completely asinine. Here are some links to relevant code areas: [MainWindow.cpp#L753](https://github.com/keepassxreboot/keepassxc/blob/develop/src/gui/MainWindow.cpp#L753) [DatabaseTabWidget.cpp#L344](https://github.com/keepassxreboot/keepassxc/blob/develop/src/gui/DatabaseTabWidget.cpp#L344)

The basic process is:

1. Program close initiated
2. If databases saved on close is on, then it connects the database closed signal to a function that stores the file of the database
3. It then issues a close all databases call which iterates a hash list to close the databases (this is where the randomness kicks in)
4. Once all are closed it saves the list of files to the ini file

**Comment by phoerious** [comment]
Yes, open databases are stored in a hash map. I stumbled over this before before when I tried to implement a selection for open databases in the auto type unlock dialog and the order of database pointers in the map had nothing to do with actual tab order. I managed to work around that, but eventually gave up, because the dialog was owned by individual database widgets, so at the time the dialog was shown, it was too late to select which database to unlock. Need to refactoring that after we merged our first KDBX4 PR.
